lib/operations/create_index.js

```diff
@@ -19,74 +20,87 @@ const keysToOmit = new Set([
+/**
+ * Creates an index on the collection.
+ *
+ * @class
+ */
+class CreateIndexOperation extends CommandOperationV2 {
+  /**
+   * Construct a CreateIndex operation.
+   *
+   * @param {Db|Collection} parent
+   * @param {string} name Collection name.
+   * @param {string|object} fieldOrSpec Defines the index.
+   * @param {object} [options] Optional settings. See Collection.prototype.createIndex for a list of options.
+   */
+  constructor(parent, name, fieldOrSpec, options) {
+    super(parent, options);
+
+    this.name = name;
 
     // Build the index
     const indexParameters = parseIndexOptions(fieldOrSpec);
     // Generate the index name
-    const indexName = typeof options.name === 'string' ? options.name : indexParameters.name;
+    this.indexName = typeof options.name === 'string' ? options.name : indexParameters.name;
     // Set up the index
-    const indexesObject = { name: indexName, key: indexParameters.fieldHash };
-
-    this.name = name;
-    this.fieldOrSpec = fieldOrSpec;
-    this.indexes = indexesObject;
-  }
-
-  _buildCommand() {
-    const options = this.options;
-    const name = this.name;
-    const indexes = this.indexes;
+    const indexes = { name: this.indexName, key: indexParameters.fieldHash };
 
     // merge all the options
     for (let optionName in options) {
       if (!keysToOmit.has(optionName)) {
         indexes[optionName] = options[optionName];
       }
     }
-
-    // Create command, apply write concern to command
-    const cmd = { createIndexes: name, indexes: [indexes] };
-
-    return cmd;
+    this.indexes = indexes;
   }
 
-  execute(callback) {
-    const db = this.db;
-    const options = this.options;
-    const indexes = this.indexes;
-
-    // Get capabilities
-    const capabilities = db.s.topology.capabilities();
-
-    // Did the user pass in a collation, check if our write server supports it
-    if (options.collation && capabilities && !capabilities.commandsTakeCollation) {
-      // Create a new error
-      const error = new MongoError('server/primary/mongos does not support collation');
-      error.code = 67;
-      // Return the error
-      return callback(error);
+  /**
+   * Execute the operation.
+   *
+   * @param {any} server
+   * @param {Collection~resultCallback} [callback] The command result callback
+   */
+  execute(server, callback) {
+    // Ensure we have a callback
+    if (this.options.writeConcern && typeof callback !== 'function') {
+      callback(
+        MongoError.create({
+          message: 'Cannot use a writeConcern without a provided callback',
+          driver: true
+        })
+      );
+      return;
     }
 
-    // Ensure we have a callback
-    if (options.writeConcern && typeof callback !== 'function') {
-      throw MongoError.create({
-        message: 'Cannot use a writeConcern without a provided callback',
-        driver: true
-      });
+    const cmd = { createIndexes: this.name, indexes: [this.indexes] };
```

Comment:
it looks like the `CreateIndexOperation` is the same as `CreateIndexesOperation` but it passes an array of one to the `createIndexes` command. Can you confirm? If that's the case, let's get rid of the singular version.

Reply:
Well, mostly. There's a difference in terms of how the `options` are handled.
 - In `createIndex`, the index options are passed into the `options` for the command. Then these options are merged into the `indexSpec` [here](https://github.com/mongodb/node-mongodb-native/pull/2326/files#diff-8385c257931c3c6f8cb44e7fa7b83f8aR49).
 - in `createIndexes`, the user just puts the options into the `indexSpec` they pass in, so there's no need to merge. The new `commitQuorum` option being the exception, which is still passed in through `options` and set at the root level of the `createIndexes` command sent to the server.

That said, it wouldn't be hard to get rid of the singular version and do the necessary merging inside the `Db`/`Collection` `createIndex` helpers.

Reply:
Alternatively, `CreateIndexOperation` could just extend `CreateIndexesOperation`.